![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ml%2Fmodels%2Fmatrix_factorization&file=matrix_factorization.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ml%2Fmodels%2Fmatrix_factorization%2Fmatrix_factorization.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/models/matrix_factorization/matrix_factorization.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Matrix Factorization — BigQuery ML

Train a **collaborative-filtering recommendation** model with `CREATE MODEL` (model_type = `MATRIX_FACTORIZATION`) — factorizes a sparse (user, item, rating) matrix into low-dimensional user and item latent factor vectors — then walk the full model lifecycle: evaluate, recommend, inspect factors, generate embeddings, item-item similarity search, introspect features/training, and tune hyperparameters.

**Lifecycle:** `CREATE MODEL` → `ML.EVALUATE` → `ML.RECOMMEND` (user→items, item→users) → `ML.WEIGHTS` → `ML.GENERATE_EMBEDDING` → `VECTOR_SEARCH` → `ML.FEATURE_INFO` → `ML.TRAINING_INFO` → hyperparameter tuning (`ML.TRIAL_INFO`)

**The one model type in this entire project that needs real infrastructure setup first:**
- **`MATRIX_FACTORIZATION` cannot train under on-demand (per-byte) pricing** — verified directly: attempting `CREATE MODEL` without a reservation fails immediately with `"Training Matrix Factorization models is not available for on-demand usage."` Every other model type in this project trains fine on-demand; this one requires a BigQuery slot reservation.
- **The reservation's edition matters too, not just its existence** — verified: a `STANDARD` edition reservation still fails, with a *different* error (`"Using BQML related functionalities is disallowed in STANDARD edition"`). `ENTERPRISE` (or higher) is required for any BQML training, matrix factorization included.
- This notebook creates a small **autoscaling** BigQuery Editions reservation (`ENTERPRISE`, 0 baseline slots, scales up to 100 only while a job actually runs) in Setup, and deletes it in Cleanup. This is **not** a capacity commitment — no upfront purchase, no minimum term; it bills per-second only for slot-time actually consumed while queries run, and idles at zero cost otherwise.
- **A newly created reservation assignment needs time to propagate** — verified: `CREATE MODEL` failed on the very first attempt right after creating and assigning a fresh reservation, and succeeded only after waiting.
- No `TRANSFORM` clause exists for this model type — the training query must produce exactly a (user, item, rating) triple.

**Data:** [`bigquery-public-data.google_analytics_sample`](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets) (Google Merchandise Store demo data, July 2017) — an **IMPLICIT** feedback signal built from product-detail-page view counts per (visitor, product) pair: 42,178 users × 320 items across ~1.3M interactions. `feedback_type='IMPLICIT'` fits this data (a behavioral proxy, not an explicit star rating) — see [RESOURCES.md](../../RESOURCES.md) for how `EXPLICIT` feedback differs (different options, different `ML.EVALUATE` metrics).

**References:** [Full reference](../../RESOURCES.md) | [CREATE MODEL (matrix factorization) docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-matrix-factorization) | [ML.RECOMMEND docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-recommend) | [Setup guide](../../setup/)

> **Cost note (measured, not estimated):** this notebook creates a temporary BigQuery Editions reservation to enable training — there's no capacity commitment and no charge while it sits idle, but training under `ENTERPRISE` autoscale bills per slot-hour of actual usage. Building and validating this notebook (base model + tuning job + the BigFrames example below, each retrained at least once) consumed **~6.4 cumulative slot-hours** of real, measured usage — check the [current BigQuery Editions pricing](https://cloud.google.com/bigquery/pricing) to estimate the actual cost for your project/region before running. The reservation is deleted in Cleanup; if a run is interrupted before Cleanup, delete it manually (see Cleanup for the exact commands) to avoid ongoing charges.

---
## Setup

Set your project and location, authenticate, and create a shared dataset for the model.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ml'  # Shared dataset across all bq-ml notebooks
RESERVATION_ID = 'bqml-mf-temp'  # Temporary reservation name (Setup creates it, Cleanup removes it)

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'google-cloud-bigquery-reservation', 'db-dtypes', 'bigquery-magics', 'bigframes', 'matplotlib')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ml ready


### Create a temporary slot reservation (required for this model type only)

`ENTERPRISE` edition, 0 baseline slots, autoscaling up to 100 slots — no capacity commitment, billed per-second only while queries actually run. Assigned to this project for `QUERY`-type jobs (BQML training runs as a `QUERY` job). Waits ~90 seconds afterward for the assignment to actually take effect — a real, verified propagation delay, not an arbitrary pause.

**Idempotent:** if this cell is re-run (e.g. after an interrupted prior run whose Cleanup never executed), it detects and reuses the existing reservation/assignment instead of erroring or creating a duplicate — and skips the 90-second wait when reusing an already-propagated assignment.

In [5]:
import time
from google.cloud import bigquery_reservation_v1
from google.api_core.exceptions import AlreadyExists

reservation_client = bigquery_reservation_v1.ReservationServiceClient()
reservation_parent = f'projects/{PROJECT_ID}/locations/{LOCATION}'
reservation_name = f'{reservation_parent}/reservations/{RESERVATION_ID}'

# Idempotent: reuse an existing reservation/assignment (e.g. from an interrupted
# prior run) instead of erroring or creating a duplicate.
try:
    reservation = reservation_client.create_reservation(
        parent=reservation_parent,
        reservation_id=RESERVATION_ID,
        reservation=bigquery_reservation_v1.Reservation(
            slot_capacity=0,
            edition=bigquery_reservation_v1.Edition.ENTERPRISE,
            autoscale=bigquery_reservation_v1.Reservation.Autoscale(max_slots=100),
            ignore_idle_slots=True,
        ),
    )
    print(f'Created {reservation.name}')
except AlreadyExists:
    reservation = reservation_client.get_reservation(name=reservation_name)
    print(f'Reservation already exists — reusing {reservation.name}')

existing_assignments = list(reservation_client.list_assignments(parent=reservation.name))
if existing_assignments:
    assignment = existing_assignments[0]
    print(f'Assignment already exists — reusing {assignment.name}')
else:
    assignment = reservation_client.create_assignment(
        parent=reservation.name,
        assignment=bigquery_reservation_v1.Assignment(
            job_type=bigquery_reservation_v1.Assignment.JobType.QUERY,
            assignee=f'projects/{PROJECT_ID}',
        ),
    )
    print(f'Created {assignment.name}, waiting ~90s for it to propagate...')
    time.sleep(90)

print('Ready to train.')

Created projects/statmike-mlops-349915/locations/US/reservations/bqml-mf-temp
Created projects/statmike-mlops-349915/locations/US/reservations/bqml-mf-temp/assignments/18001908066168517460, waiting ~90s for it to propagate...
Ready to train.


---
## Step 1 — Create the model with `CREATE MODEL`

`feedback_type = 'IMPLICIT'` fits this data — `view_count` is a behavioral proxy signal, not an explicit star rating. `num_factors = 16` is a deliberate, modest choice for a 320-item catalog. There is no `TRANSFORM` clause for this model type — the training `SELECT` must produce exactly a (user, item, rating) triple.

In [6]:
query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`
OPTIONS(
  model_type = 'MATRIX_FACTORIZATION',
  feedback_type = 'IMPLICIT',
  user_col = 'visitor_id',
  item_col = 'product_name',
  rating_col = 'view_count',
  num_factors = 16
) AS
SELECT
  fullVisitorId AS visitor_id,
  product.v2ProductName AS product_name,
  COUNT(*) AS view_count
FROM `bigquery-public-data.google_analytics_sample.ga_sessions_*`,
UNNEST(hits) AS hits,
UNNEST(hits.product) AS product
WHERE _TABLE_SUFFIX BETWEEN '20170701' AND '20170801'
  AND product.v2ProductName IS NOT NULL AND product.v2ProductName != '(not set)'
GROUP BY visitor_id, product_name
"""
client.query(query).result()
print('Model matrix_factorization_ga created')

Model matrix_factorization_ga created


---
## Step 2 — Evaluate with `ML.EVALUATE`

IMPLICIT feedback gets ranking-style metrics — `mean_average_precision` and `normalized_discounted_cumulative_gain` (both higher-is-better, bounded [0,1]), plus `mean_squared_error` and `average_rank`. `EXPLICIT` models get a different, regression-style metric set instead (see [RESOURCES.md](../../RESOURCES.md)) — the two feedback types aren't evaluated the same way.

> **Verified:** retraining this exact model (same name, same SQL) does not reproduce `mean_average_precision` exactly — one training run reached `0.873`, another reached `0.860`. Matrix factorization's WALS training shows measurable retraining variance, similar to [K-Means](../kmeans/)/`RANDOM_FOREST_*` elsewhere in this project, unlike PCA's full determinism or the DNN family's bit-for-bit reproducibility under a fixed name.

In [7]:
query = f"""
SELECT *
FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`)
"""
client.query(query).to_dataframe()

,mean_average_precision,mean_squared_error,normalized_discounted_cumulative_gain,average_rank
0,0.86022,0.06475,0.941836,0.479667


---
## Step 3 — Recommend items for a user with `ML.RECOMMEND`

Passing only the user column scores every item for that user.

> **GOTCHA (verified):** calling this for a `visitor_id` **not present** in the training data does **not** error — it silently returns a full ranked list. [RESOURCES.md](../../RESOURCES.md)'s general cold-start note ("cannot recommend for users or items absent from training data") describes the intent, not the observed runtime behavior: comparing the output for a confirmed real trained user against an obviously fake user ID shows completely different top-5 lists, consistent with the fake user falling back to item-side bias/popularity alone rather than genuine personalization (no user factor was ever learned for it). Don't assume a non-empty `ML.RECOMMEND` result means the input user actually existed in training.

In [8]:
query = f"""
SELECT visitor_id, product_name, predicted_view_count_confidence
FROM ML.RECOMMEND(
  MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`,
  (SELECT '9790296909741758431' AS visitor_id)
)
ORDER BY predicted_view_count_confidence DESC
LIMIT 5
"""
client.query(query).to_dataframe()

,visitor_id,product_name,predicted_view_count_confidence
0,9790296909741758431,Windup Android,0.910649
1,9790296909741758431,Foam Can and Bottle Cooler,0.908474
2,9790296909741758431,Android Luggage Tag,0.899293
3,9790296909741758431,Android Lunch Kit,0.893223
4,9790296909741758431,Google Luggage Tag,0.885585


---
## Step 4 — Recommend users for an item with `ML.RECOMMEND`

Passing only the item column scores every user for that item — the symmetric counterpart to Step 3.

In [9]:
query = f"""
SELECT visitor_id, product_name, predicted_view_count_confidence
FROM ML.RECOMMEND(
  MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`,
  (SELECT 'Google Youth Girl Hoodie' AS product_name)
)
ORDER BY predicted_view_count_confidence DESC
LIMIT 5
"""
client.query(query).to_dataframe()

,visitor_id,product_name,predicted_view_count_confidence
0,0878293552935562825,Google Youth Girl Hoodie,1.161636
1,5938467158077430288,Google Youth Girl Hoodie,1.100342
2,5735858296023788680,Google Youth Girl Hoodie,1.079091
3,521290356169916504,Google Youth Girl Hoodie,1.078686
4,6481734531894360944,Google Youth Girl Hoodie,1.066481


---
## Step 5 — Inspect the learned factors with `ML.WEIGHTS`

One row per user AND per item (distinguished by `processed_input`), each with a `factor_weights` `ARRAY<STRUCT<factor, weight>>` (length = `num_factors`) and an `intercept` (per-user/per-item bias term).

> **GOTCHA (verified):** there is also exactly one extra row with `processed_input=NULL` and `feature='global__INTERCEPT__'` — the single global bias term shared across the whole model, separate from every user's/item's own intercept.

In [10]:
query = f"""
SELECT processed_input, COUNT(*) AS n
FROM ML.WEIGHTS(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`)
GROUP BY processed_input
"""
client.query(query).to_dataframe()

,processed_input,n
0,visitor_id,42178
1,None,1
2,product_name,320


---
## Step 6 — Factor vectors as embeddings with `ML.GENERATE_EMBEDDING`

> **GOTCHA (verified):** for `MATRIX_FACTORIZATION`, `ML.GENERATE_EMBEDDING` takes **only the model** — no input table/query argument at all. Passing one errors immediately: `"Function ML.GENERATE_EMBEDDING for MATRIX_FACTORIZATION models only expects 1 argument but 2 were passed."` This is a genuinely different signature from [PCA](../pca/)/[Autoencoder](../autoencoder/) (both require a 2nd input-data argument) — it returns embeddings for **every** user and item in the model in one call, mirroring `ML.WEIGHTS`' `processed_input`/`feature` shape rather than `ML.PREDICT`'s.
>
> **Second GOTCHA (verified):** the embedding array has **`num_factors + 1` elements (17 here, not 16)** — one more than `ML.WEIGHTS`' `factor_weights` array for the same model (confirmed 16 elements there, matching `num_factors` exactly). The extra element is almost certainly the per-user/per-item `intercept` appended onto the raw factor vector. Don't assume the embedding array length always equals `num_factors` for this model type.

In [11]:
query = f"""
SELECT processed_input, feature, ml_generate_embedding_result
FROM ML.GENERATE_EMBEDDING(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`)
LIMIT 5
"""
client.query(query).to_dataframe()

,processed_input,feature,ml_generate_embedding_result
0,visitor_id,30435529137043710,"[4.892840428765869, 1.2575409884758202, 3.245236910788326, -4.2408612647078385, 2.5235300470413975, 0.8711630838005551, 10.54365113330722, -8.581271607479813, -4.676225245315781, 5.888235162245928, 3.0224187236361466, 3.1183571121834532, 5.374350256172026, -7.800331396409579, -1.5958897554806522, 1.0448244616804891, -3.4315091132431825]"
1,visitor_id,092900555524137497,"[4.486530258206768, 3.128540106796451, 1.8176944020826187, -2.6326362387502993, 0.3234151691607221, 3.798659648653218, 8.655098232290902, -7.207356818123239, -2.152311000711968, 6.42357468060258, 4.708260451159926, 4.81973277592189, 5.741736134455737, -2.5347267030085376, -3.9229074454083688, 1.2187962546315323, -3.3877608699722623]"
2,visitor_id,5510388986800068083,"[-0.8054126780823315, 5.7034857190403745, 2.116391901622975, -0.9696384553172338, 0.4280694074179656, 5.033011005383364, 8.667358436729442, -10.653391379218249, -4.630745843023025, 6.18518414869284, 6.616911191477075, 2.249596844655178, 2.9952112798638595, -3.538194838280644, -3.2762179637766824, 2.3087321948201875, -3.290870819885016]"
3,visitor_id,1835601435265620139,"[7.294203814262676, -6.228393070533617, 5.0633896154499585, -3.493717966796081, -3.3739144950540663, -1.880584565383588, 6.611177391932552, -7.284582070038838, -12.155797046042261, 4.206500364853954, -0.8202270029062089, 5.35085968103206, 0.42125526608821645, -3.7766018384684825, -8.873191603461487, -4.550422211361446, -3.200110846859248]"
4,visitor_id,6193351678286788918,"[7.550779278841895, -1.8053944850569366, 0.8765240220812602, -5.61343028389748, -0.6616194809009583, -1.4487357370730702, 9.283411180863835, -5.034655813231021, -3.4836433902225186, 4.66431511617183, -0.23582970087159838, 6.27002830836186, 6.412685793614576, -5.299144851286753, -4.961178845448039, -2.212289637948431, -3.1053261191981423]"


---
## Step 7 — Item-item similarity with `VECTOR_SEARCH`

Materialize the item embeddings (Step 6's output filtered to `processed_input='product_name'`) into a real table — `VECTOR_SEARCH` doesn't accept `ML.GENERATE_EMBEDDING` output directly as its base-table argument, the same limitation already verified in [Autoencoder](../autoencoder/).

In [12]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_item_embeddings` AS
SELECT feature AS product_name, ml_generate_embedding_result AS embedding
FROM ML.GENERATE_EMBEDDING(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`)
WHERE processed_input = 'product_name'
"""
client.query(query).result()
print('Table matrix_factorization_item_embeddings created')

Table matrix_factorization_item_embeddings created


> **Verified:** querying "Google Youth Girl Hoodie" surfaces other youth apparel (youth tees, youth t-shirts) as its nearest neighbors — the learned item factors capture real product-category structure, not just noise.

In [13]:
query = f"""
SELECT query.product_name AS query_product, base.product_name AS similar_product, distance
FROM VECTOR_SEARCH(
  TABLE `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_item_embeddings`, 'embedding',
  (SELECT product_name, embedding
   FROM `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_item_embeddings`
   WHERE product_name = 'Google Youth Girl Hoodie'),
  top_k => 6,
  distance_type => 'COSINE'
)
ORDER BY distance
"""
client.query(query).to_dataframe()

,query_product,similar_product,distance
0,Google Youth Girl Hoodie,Google Youth Girl Hoodie,0.000000
1,Google Youth Girl Hoodie,Android Youth Short Sleeve T-shirt Pewter,0.000251
2,Google Youth Girl Hoodie,Android Youth Short Sleeve T-shirt Aqua,0.000270
3,Google Youth Girl Hoodie,Google Youth Short Sleeve T-shirt Yellow,0.000295
4,Google Youth Girl Hoodie,Google Youth Short Sleeve T-shirt Green,0.000298
5,Google Youth Girl Hoodie,Google Youth Sports Tee Navy,0.000331


---
## Step 8 — Introspect the training columns with `ML.FEATURE_INFO`


In [14]:
query = f"""
SELECT *
FROM ML.FEATURE_INFO(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`)
"""
client.query(query).to_dataframe()

,input,min,max,mean,median,stddev,category_count,null_count,dimension
0,view_count,1.0,298.0,1.557347,1.0,1.740846,<NA>,0,<NA>
1,visitor_id,NaN,NaN,NaN,NaN,NaN,42178,0,<NA>
2,product_name,NaN,NaN,NaN,NaN,NaN,320,0,<NA>


---
## Step 9 — Training loss curve with `ML.TRAINING_INFO`

> **Verified:** `eval_loss` is `NULL` throughout for this model — unlike the supervised model types in this project, `MATRIX_FACTORIZATION`'s `ML.TRAINING_INFO` doesn't populate a held-out `eval_loss` column here.

In [15]:
query = f"""
SELECT *
FROM ML.TRAINING_INFO(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga`)
"""
client.query(query).to_dataframe()

,training_run,iteration,loss,eval_loss,duration_ms
0,0,6,0.017670,NaN,33345
1,0,5,0.018472,NaN,33082
2,0,4,0.019612,NaN,32647
3,0,3,0.021799,NaN,29230
4,0,2,0.028137,NaN,26241
5,0,1,0.046729,NaN,34367
6,0,0,144.525843,NaN,53228


---
## Step 10 — Hyperparameter tuning

Tunes `num_factors`, `l2_reg`, and `wals_alpha` (IMPLICIT-only) together, maximizing `mean_average_precision` (the IMPLICIT default objective).

> **Verified:** tuning improved on Step 2's baseline here — the best trial (`num_factors=26`, `l2_reg=2.88`, `wals_alpha=49.2`) reached `mean_average_precision=0.899`, versus this run's untuned baseline of `0.860`. A separate training of the identical baseline and tuning config reached different specific numbers (baseline `0.873`, best tuned trial `0.915` with `wals_alpha=30.8` instead) — consistent with Step 2's retraining-variance finding, the exact hyperparameter search results aren't bit-for-bit reproducible either. **The qualitative result held in both runs: tuning beat the untuned baseline.** Contrast with [Autoencoder](../autoencoder/), where an equally-sized 4-trial search failed to beat its untuned baseline at all.

In [16]:
query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga_tuned`
OPTIONS(
  model_type = 'MATRIX_FACTORIZATION',
  feedback_type = 'IMPLICIT',
  user_col = 'visitor_id',
  item_col = 'product_name',
  rating_col = 'view_count',
  num_factors = HPARAM_RANGE(8, 32),
  l2_reg = HPARAM_RANGE(0.1, 10.0),
  wals_alpha = HPARAM_RANGE(20.0, 60.0),
  num_trials = 4,
  max_parallel_trials = 2
) AS
SELECT
  fullVisitorId AS visitor_id,
  product.v2ProductName AS product_name,
  COUNT(*) AS view_count
FROM `bigquery-public-data.google_analytics_sample.ga_sessions_*`,
UNNEST(hits) AS hits,
UNNEST(hits.product) AS product
WHERE _TABLE_SUFFIX BETWEEN '20170701' AND '20170801'
  AND product.v2ProductName IS NOT NULL AND product.v2ProductName != '(not set)'
GROUP BY visitor_id, product_name
"""
client.query(query).result()
print('Tuned model created')

Tuned model created


In [17]:
query = f"""
SELECT
  trial_id,
  hyperparameters,
  hparam_tuning_evaluation_metrics.mean_average_precision AS mean_average_precision,
  is_optimal
FROM ML.TRIAL_INFO(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga_tuned`)
ORDER BY mean_average_precision DESC
"""
client.query(query).to_dataframe()

,trial_id,hyperparameters,mean_average_precision,is_optimal
0,3,"{'l2_reg': 2.8822008689901706, 'num_factors': 26, 'wals_alpha': 49.197708975742735}",0.899370,True
1,1,"{'l2_reg': 1.0000000000000002, 'num_factors': 20, 'wals_alpha': 40.0}",0.888839,False
2,4,"{'l2_reg': 8.900627435101063, 'num_factors': 20, 'wals_alpha': 39.430175685775424}",0.872179,False
3,2,"{'l2_reg': 0.3716244478653518, 'num_factors': 14, 'wals_alpha': 31.436637102137972}",0.861319,False


Calling `ML.EVALUATE` on a tuned model with no extra arguments returns every trial's metrics in one result (one row per `trial_id`) — there's no `STRUCT(trial_id)` 2-argument form for `ML.EVALUATE` the way `ML.RECOMMEND`/`ML.PREDICT` accept one; passing a `STRUCT` as a 2nd argument errors (`"argument 2 must be a relation"`).

In [18]:
query = f"""
SELECT *
FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_ga_tuned`)
"""
client.query(query).to_dataframe()

,trial_id,mean_average_precision,mean_squared_error,normalized_discounted_cumulative_gain,average_rank
0,1,0.888839,0.056201,0.941149,0.479083
1,2,0.861319,0.064673,0.943073,0.479141
2,3,0.899370,0.050711,0.940095,0.479505
3,4,0.872179,0.063256,0.943588,0.477650


---
## Examples — `%%bigquery` Magics

The same operations using IPython magic commands — write SQL directly in cells without Python string wrapping.

- `%%bigquery` — run SQL, display results inline
- `%%bigquery df` — run SQL, capture results into a pandas DataFrame

### Evaluate with `%%bigquery`

In [19]:
%%bigquery --project {PROJECT_ID}

SELECT *
FROM ML.EVALUATE(MODEL `statmike-mlops-349915.bq_ml.matrix_factorization_ga`)

Query is running:   0%|          |

Downloading:   0%|          |

,mean_average_precision,mean_squared_error,normalized_discounted_cumulative_gain,average_rank
0,0.86022,0.06475,0.941836,0.479667


---
## Examples — BigFrames

BigFrames provides a scikit-learn-style API (`bigframes.ml`) that trains BigQuery ML models under the hood. **Matrix factorization has a genuine first-class wrapper: `bigframes.ml.decomposition.MatrixFactorization`.**

> **GOTCHA (verified against the live installed constructor signature):** `MatrixFactorization(feedback_type=..., num_factors=..., user_col=..., item_col=..., rating_col=..., l2_reg=...)` — there is **no `wals_alpha` parameter**, unlike the SQL `CREATE MODEL` syntax. For `feedback_type='implicit'`, BigFrames trains with BigQuery ML's default `wals_alpha` (40) with no way to override it through this wrapper — use SQL directly if you need to set `wals_alpha`.

In [20]:
import bigframes.pandas as bpd

bpd.close_session()  # Reset session to apply project/location settings
bpd.options.bigquery.project = PROJECT_ID
bpd.options.bigquery.location = LOCATION

In [21]:
from bigframes.ml.decomposition import MatrixFactorization

query = f"""
SELECT
  fullVisitorId AS visitor_id,
  product.v2ProductName AS product_name,
  COUNT(*) AS view_count
FROM `bigquery-public-data.google_analytics_sample.ga_sessions_*`,
UNNEST(hits) AS hits,
UNNEST(hits.product) AS product
WHERE _TABLE_SUFFIX BETWEEN '20170701' AND '20170801'
  AND product.v2ProductName IS NOT NULL AND product.v2ProductName != '(not set)'
GROUP BY visitor_id, product_name
"""
df = bpd.read_gbq(query)

# Train (creates a BigQuery ML model behind the scenes)
model = MatrixFactorization(
    feedback_type='implicit',
    num_factors=16,
    user_col='visitor_id',
    item_col='product_name',
    rating_col='view_count',
)
model.fit(df)

# Evaluate
model.score(df).to_pandas()

,mean_average_precision,mean_squared_error,normalized_discounted_cumulative_gain,average_rank
0,0.860575,0.065071,0.941422,0.479598


---
## Cleanup

Drop the models and table created by this notebook, then delete the temporary reservation and its assignment. The shared dataset is left for other notebooks.

In [22]:
client.query(f'DROP TABLE IF EXISTS `{PROJECT_ID}.{DATASET_ID}.matrix_factorization_item_embeddings`').result()
print('Dropped matrix_factorization_item_embeddings')

for model_name in [
    'matrix_factorization_ga',
    'matrix_factorization_ga_tuned',
]:
    client.query(f'DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.{model_name}`').result()
    print(f'Dropped {model_name}')

Dropped matrix_factorization_item_embeddings
Dropped matrix_factorization_ga
Dropped matrix_factorization_ga_tuned


In [23]:
# Delete the temporary reservation and its assignment (safe to re-run — ignores NotFound)
from google.api_core.exceptions import NotFound

try:
    for existing_assignment in reservation_client.list_assignments(parent=reservation.name):
        reservation_client.delete_assignment(name=existing_assignment.name)
    reservation_client.delete_reservation(name=reservation.name)
    print(f'Deleted reservation {reservation.name} and its assignment(s)')
except NotFound:
    print('Reservation already deleted')

Deleted reservation projects/statmike-mlops-349915/locations/US/reservations/bqml-mf-temp and its assignment(s)


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and everything in it.

In [24]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')